### PR TITLE
Performance fixes for contact history

### PR DESCRIPTION
**Pull request by ericnewcomer**
Two changes to deal with timeouts for some contacts
* Truncate the number of items fetched per page regardless of time window (one example contact had many thousands of activity items in a 90 day window)
* Add contact to WebHookResult to avoid join on runs table

### CODE REVIEW

File: temba/api/migrations/0015_webhookresult_contact.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        ('api', '0014_auto_20170410_0731'),
+    ]
+
+    operations = [
+
+        # add nullable contact field
+        migrations.AddField(
+            model_name='webhookresult',
+            name='contact',
+            field=models.ForeignKey(help_text='The contact that generated this result', null=True, on_delete=django.db.models.deletion.CASCADE, related_name='results', to='contacts.Contact'),
```
Comment (rowanseymour) [inline temba/api/migrations/0015_webhookresult_contact.py]: The related_name could be called `webhook_results` to clarify it's not flow results etc.

File: temba/api/migrations/0015_webhookresult_contact.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.2 on 2017-10-11 20:46
+from __future__ import unicode_literals
+
+from django.db import migrations, models
+import django.db.models.deletion
+
+
+def backfill_webhook_result_contact(apps, schema_editor):
```
Comment (rowanseymour) [inline temba/api/migrations/0015_webhookresult_contact.py]: Not sure how much we're worried about performance here, but we could:
1. Fetch results in batches with prefetch for `event__run__contact`
2. Only prefetch event__run and just copy over the contact_id value
3. Use update_fields in result.save()
4. Possibly update results in batches using an F() expression to define the contact id value

File: temba/api/migrations/0015_webhookresult_contact.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    WebHookResult = apps.get_model('api', 'WebHookResult')
+    for result in WebHookResult.objects.filter(contact=None).select_related('event__run__contact'):
```
Comment (nicpottier) [inline temba/api/migrations/0015_webhookresult_contact.py]: Don't we only want to do WebHookResults which have an associated event? (we may have results for the org level webhooks)

File: temba/contacts/models.py [hunk 4]
```diff
@@ -582,37 +583,41 @@ def get_activity(self, after, before):
...
 
         # and all of this contact's runs, channel events such as missed calls, scheduled events
-        runs = self.runs.filter(created_on__gte=after, created_on__lt=before).exclude(flow__flow_type=Flow.MESSAGE)
-        runs = runs.select_related('flow')
+        started_runs = self.runs.filter(created_on__gte=after, created_on__lt=before).exclude(flow__flow_type=Flow.MESSAGE)
+        started_runs = started_runs.order_by('-created_on').select_related('flow')[:MAX_HISTORY]
+
+        exited_runs = self.runs.filter(exited_on__gte=after, exited_on__lt=before).exclude(flow__flow_type=Flow.MESSAGE)
+        exited_runs = exited_runs.exclude(exit_type=None).order_by('-created_on').select_related('flow')[:MAX_HISTORY]
+        print (str(exited_runs.query))
```
Comment (nicpottier) [inline temba/contacts/models.py]: !!

File: temba/contacts/models.py [hunk 5]
```diff
@@ -582,37 +583,41 @@ def get_activity(self, after, before):
...
+            [{'type': 'run-exit', 'time': r.exited_on, 'obj': r} for r in exited_runs],
             [{'type': 'channel-event', 'time': e.created_on, 'obj': e} for e in channel_events],
             [{'type': 'event-fire', 'time': f.fired, 'obj': f} for f in event_fires],
-            [{'type': 'webhook-result', 'time': r.created_on, 'obj': r} for r in webhook_result],
+            [{'type': 'webhook-result', 'time': r.created_on, 'obj': r} for r in webhook_results],
             [{'type': 'call', 'time': c.created_on, 'obj': c} for c in calls],
         )
 
-        return sorted(activity, key=lambda i: i['time'], reverse=True)
+        return sorted(activity, key=lambda i: i['time'], reverse=True)[:MAX_HISTORY]
```
Comment (nicpottier) [inline temba/contacts/models.py]: Is this right? Don't you need to trim all of these lists to the latest based on `max_history`? IE, if there are 500 messages and you only grab most recent 50, you only want to display webhook results that are within the first 50. As this is won't we show 50 webhook results which may be from earlier than the 50 messages we are displaying?

Comment (nicpottier) [reply temba/contacts/models.py]: NM, this is rigth! great success!

**Pull request merged**